### PR TITLE
feat(kumactl) add install loki for log aggregation

### LINKED ISSUE #819: Add log aggregation  to kumactl install
### Summary

The goal for `kumactl install` to provide a way to aggregate and visualize app logs easily.  
I propose to use [grafana/Loki](https://github.com/grafana/loki) to aggregate app logs and visualize then directly into the Grafana installed with the metrics. 

This screen shot shows how log visualization with Loki can be used with the traces from Jaeger, if your logs contain the traceId. Some libraries add the traceId into the logs automatically (e.g. [Sleuth](https://github.com/spring-cloud/spring-cloud-sleuth), for Spring java apps).

Capture d’écran 2020-06-10 à 10 12 05

### PR DESCRIPTION
**Pull request by xbauquet**
### Summary

Add the Loki stack to `kumactl install` allowing to see logs
directly in Grafana. The yaml for loki was inspired by:
`https://grafana.github.io/loki/charts`

### Full changelog

* Add a new install command to kubectl:
  `kubectl install logging`
* Upgrate Grafana version to 7.0.3 that accepts loki as a
  data source.

### Issues resolved

#819

### Documentation

Documentation PR: Kong/kuma-website#222

### CODE REVIEW

**Comment by xbauquet** [comment]
Follow up on the meeting of the 2020/06/10:

## Question 
Can we use Loki and Grafana to aggregate and visualise TrafficLogs ? 

## Response
Here are the steps of my test: 

### Step 1: Access logs in default format

My first try was to make the data-planes send the Access Logs to `stdout`
 
```
logging:
    defaultBackend: file
    backends:
      - name: file
        type: file
        conf:
          path: /dev/stdout
```

That worked well I get the Access Logs with the default format.

Capture d’écran 2020-06-11 à 17 19 32

### Step 2: Access logs in custom format

My second test was to use a custom format for my Access Logs
```
logging:
    defaultBackend: file
    backends:
      - name: file
        type: file
        format: '{"start_time": "%START_TIME%", "source": "%KUMA_SOURCE_SERVICE%", "destination": "%KUMA_DESTINATION_SERVICE%"}'
        conf:
          path: /dev/stdout
```

When I use the `format` I loose the Access Logs

### Step 3: Back to the default format

This is where this test becomes weird. 
I removed the custom format to be back with the configuration in Step 1. 
As a result my Access Logs are back in default format BUT my Access Logs with the custom format from the Step 2 are displayed concatenated in one single line.

Capture d’écran 2020-06-11 à 17 24 19

### Conclusion 
 
It looks like using a custom format with the `type: file` makes the data-planes concatenate all the Access Logs until a new Log is generated.
I don't know where to look for that specific problem, but this could be a little bug. 
If someone has any idea ?!!

**Comment by jakubdyszkiewicz** [comment]
Try to put the new line in the format. Something like this:
```
logging:
    defaultBackend: file
    backends:
      - name: file
        type: file
        format: '{"start_time": "%START_TIME%", "source": "%KUMA_SOURCE_SERVICE%", "destination": "%KUMA_DESTINATION_SERVICE%"}
'
        conf:
          path: /dev/stdout
```
or
```
logging:
    defaultBackend: file
    backends:
      - name: file
        type: file
        format: | {"start_time": "%START_TIME%", "source": "%KUMA_SOURCE_SERVICE%", "destination": "%KUMA_DESTINATION_SERVICE%"}

        conf:
          path: /dev/stdout
```

**Comment by xbauquet** [comment]
I'll try that thank you.

**Comment by xbauquet** [comment]
```
  logging:
    defaultBackend: file
    backends:
      - name: file
        format: |
          {"start_time": "%START_TIME%", "source": "%KUMA_SOURCE_SERVICE%", "destination": "%KUMA_DESTINATION_SERVICE%"}

        type: file
        conf:
          path: /dev/stdout
```

This work (not the other option).
This is a good work around but I still think we should fix it when possible.

**Comment by jakubdyszkiewicz** [comment]
To be honest, I'll be willing to add the new line automatically in `http_access_log_configurer` and in `network_access_log_configurer`.

I see we do it in default format. I'd suggest a change so instead of doing it only for default format, we add it for all formats, even ones provided by the user.

**Comment by xbauquet** [comment]
I think adding automatically a new line for all formats is a good idea.
Can you tell me where is the code taking care of that? I can do the modification if you want.

**Comment by jakubdyszkiewicz** [comment]
I think `convertLoggingBackend` is a place. Add a new line before
`format, err := accesslog.ParseFormat(formatString)`

Also I think that in result of this change, the line
`kuma/app/kuma-dp/pkg/dataplane/accesslogs/sender.go:33` is not needed anymore.

**Comment by tharun208** [comment]
@xbauquet , Imports are not ordered. try to run `make imports`.

**Comment by xbauquet** [comment]
Documentation PR: https://github.com/Kong/kuma-website/pull/222

File: app/kuma-dp/pkg/dataplane/accesslogs/sender.go [hunk 1]
```diff
@@ -30,7 +30,7 @@ func (s *sender) Connect() error {
 }
 
 func (s *sender) Send(record string) error {
-	_, err := s.conn.Write(append([]byte(record), byte('\n')))
+	_, err := s.conn.Write([]byte(record))
```
Comment (nickolaev) [inline app/kuma-dp/pkg/dataplane/accesslogs/sender.go:33]: @jakubdyszkiewicz what do you think?

**Comment by xbauquet** [comment]
I just looked at the Loki documentation and it is possible to install it without Kubernetes so I guess I could do a bit of doc on how to install it for universal or at least add some links to help people installing it.

**Comment by nickolaev** [comment]
Can you also:
 - merge master
 - run `make check`, `make test`, `make integration` and `make test/e2e` (all separate as some of these are pretty heavy). The latter two assume you have Docker running and accissebal from your account. Please make sure you fix tests related to logging. If something else fails, reach out over slack to check it with us.

Thanks a lot for your contributions!

**Comment by xbauquet** [comment]
I opened a issue #841 because I have a problem when trying to fix tests.

**Comment by xbauquet** [comment]
@nickolaev I corrected all the tests I could. I still see a few problems in Circle CI but I don't understand the problem :s

**Comment by nickolaev** [comment]
@xbauquet CI is sometimes flaky. I re-run the failing tests and all is fine now

File: app/kumactl/data/install/k8s/metrics/grafana/grafana.yaml [hunk 2]
```diff
@@ -263,7 +263,7 @@ spec:
         runAsUser: 472
       containers:
         - name: grafana
-          image: "grafana/grafana:6.6.0"
+          image: "grafana/grafana:7.0.3"
```
Comment (nickolaev) [inline app/kumactl/data/install/k8s/metrics/grafana/grafana.yaml:266]: Please bump the version to 7.0.5.

**Pull request merged**